lib/exercises.rb

```diff
@@ -1,19 +1,41 @@
 
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n * m) n is the length of the array, m is the max length of string within the array. 
+# Could pull chars method to run first on each element of strings, for O(n) n being the longest string or array.
+# Space Complexity: O(n)
 
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  return [] if strings.nil?
+  return [] if strings.empty?
+
+  counts_hash = Hash.new([])
+  strings.each do |string|
+    counts_hash[string.chars.sort] += [string]
+  end
+  return counts_hash.map {|k, v| v}
 end
 
 # This method will return the k most common elements
 # in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n) n is list length
+# Space Complexity: O(n^2 * k) is that right?
 def top_k_frequent_elements(list, k)
```

Comment:
👍 The time complexity is O(n log n) because you have an O(n) loop (the `list.each` loop) then a sort (O(n log n)), and followed by another loop which goes k times where k <= n.

So the time complexity is O(n + n log n + k)  ==> O(n log n) because `n log n` is the dominate term.  It's larger than n and k.

In the space complexity you're creating a hash O(n), a sorted hash O(n), and `k_most_common_ele`.  So O(n + n  + k) and so the final space complexity is O(n).